Wrong raft messages may cause etcd panic

### Bug report criteria

- [X] This bug report is not security related, security issues should be disclosed privately via security@etcd.io.
- [X] This is not a support request or question, support requests or questions should be raised in the etcd [discussion forums](https://github.com/etcd-io/etcd/discussions).
- [X] You have read the etcd [bug reporting guidelines](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/reporting_bugs.md).
- [X] Existing open issues along with etcd [frequently asked questions](https://etcd.io/docs/latest/faq) have been checked and this is not a duplicate.

### What happened?

Multiple etcd servers repeatedly panic with a message "tocommit(4432450) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"

The first panic occurred one month after the etcd started running. At that time, there were some network issues, leading to TCP reconnections.




### What did you expect to happen?

No panic.

### How can we reproduce it (as minimally and precisely as possible)?

It's hard to reproduce the panic in real world, but we have reproduced it in a unit test.

### Anything else we need to know?

After investigation, we suspect that a malfunctioning switch or a similar issue may be dispatching incorrect Raft messages, leading to the panic. Currently, etcd does not check the `To` field in messages.


### Etcd version (please run commands below)

<details>

We ([tikv/pd](https://github.com/tikv/pd/)) use embed etcd v3.4.21 https://github.com/etcd-io/etcd/commit/85b640cee793

</details>


### Etcd configuration (command line flags or environment variables)

<details>

Default settings. 

</details>


### Etcd debug information (please run commands below, feel free to obfuscate the IP address or FQDN in the output)

<details>

```console
$ etcdctl member list -w table
# paste output here

$ etcdctl --endpoints=<member list> endpoint status -w table
# paste output here
```

</details>


### Relevant log output

```Shell
[2023/10/30 14:48:34.402 +08:00] [WARN] [stream.go:266] ["closed TCP streaming connection with remote peer"] [stream-writer-type="stream MsgApp v2"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=18cf5532a87c6527]
[2023/10/30 14:48:34.402 +08:00] [WARN] [stream.go:277] ["established TCP streaming connection with remote peer"] [stream-writer-type="stream MsgApp v2"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=18cf5532a87c6527]
[2023/10/30 14:48:34.409 +08:00] [INFO] [stream.go:425] ["established TCP streaming connection with remote peer"] [stream-reader-type="stream MsgApp v2"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=18cf5532a87c6527]
[2023/10/30 14:48:34.409 +08:00] [INFO] [stream.go:425] ["established TCP streaming connection with remote peer"] [stream-reader-type="stream MsgApp v2"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=1baed0b840546f7]
[2023/10/30 14:48:34.410 +08:00] [INFO] [stream.go:425] ["established TCP streaming connection with remote peer"] [stream-reader-type="stream Message"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=18cf5532a87c6527]
[2023/10/30 14:48:34.410 +08:00] [INFO] [stream.go:425] ["established TCP streaming connection with remote peer"] [stream-reader-type="stream Message"] [local-member-id=d4c1cc21d4c817ca] [remote-peer-id=1baed0b840546f7]
[2023/10/30 14:48:34.419 +08:00] [PANIC] [log.go:203] ["tocommit(4432450) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]
...
[2023/10/30 14:48:34.419 +08:00] [PANIC] [log.go:203] ["tocommit(4432450) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]
...
[2023/10/30 14:48:54.770 +08:00] [PANIC] [log.go:203] ["tocommit(4432459) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]
...
[2023/10/30 14:49:13.070 +08:00] [PANIC] [log.go:203] ["tocommit(4432492) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]
...
[2023/10/30 14:49:31.681 +08:00] [PANIC] [log.go:203] ["tocommit(4432521) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]
...
[2023/10/30 14:49:49.816 +08:00] [PANIC] [log.go:203] ["tocommit(4432522) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?"]


panic: tocommit(4432450) is out of range [lastIndex(4432444)]. Was the raft log corrupted, truncated, or lost?

goroutine 249 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc00456e9c0, {0x0, 0x0, 0x0})
        /go/pkg/mod/go.uber.org/zap@v1.19.1/zapcore/entry.go:232 +0x44c
go.uber.org/zap.(*SugaredLogger).log(0xc0002bad50, 0x4, {0x2b11764?, 0xdcbc85?}, {0xc007dac240?, 0x26ef300?, 0xc0098aa300?}, {0x0, 0x0, 0x0})
        /go/pkg/mod/go.uber.org/zap@v1.19.1/sugar.go:227 +0xee
go.uber.org/zap.(*SugaredLogger).Panicf(...)
        /go/pkg/mod/go.uber.org/zap@v1.19.1/sugar.go:159
go.etcd.io/etcd/pkg/logutil.(*zapRaftLogger).Panicf(0x439fec?, {0x2b11764?, 0x2b?}, {0xc007dac240?, 0x1672334?, 0x3cd6081?})
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/pkg/logutil/zap_raft.go:101 +0x49
go.etcd.io/etcd/raft.(*raftLog).commitTo(0xc000760150, 0x43a001)
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/log.go:203 +0x103
go.etcd.io/etcd/raft.(*raft).handleHeartbeat(_, {0x8, 0x18cf5532a87c6527, 0xd4c1cc21d4c817ca, 0x18, 0x0, 0x0, {0x0, 0x0, 0x0}, ...})
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/raft.go:1389 +0x45
go.etcd.io/etcd/raft.stepCandidate(_, {0x8, 0x18cf5532a87c6527, 0xd4c1cc21d4c817ca, 0x18, 0x0, 0x0, {0x0, 0x0, 0x0}, ...})
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/raft.go:1289 +0x585
go.etcd.io/etcd/raft.(*raft).Step(_, {0x8, 0x18cf5532a87c6527, 0xd4c1cc21d4c817ca, 0x18, 0x0, 0x0, {0x0, 0x0, 0x0}, ...})
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/raft.go:990 +0x1055
go.etcd.io/etcd/raft.(*node).run(0xc000118d80)
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/node.go:354 +0xa05
created by go.etcd.io/etcd/raft.RestartNode
        /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/raft/node.go:242 +0x24a
```

--- a/server/etcdserver/server.go
+++ b/server/etcdserver/server.go
@@ -698,6 +698,14 @@ func (s *EtcdServer) Process(ctx context.Context, m raftpb.Message) error {
 		)
 		return httptypes.NewHTTPError(http.StatusForbidden, "cannot process message from removed member")
 	}
+	if s.MemberId() != types.ID(m.To) {
+		lg.Warn(
+			"rejected Raft message to mismatch member",
+			zap.String("local-member-id", s.MemberId().String()),
+			zap.String("mismatch-member-id", types.ID(m.To).String()),
+		)
+		return httptypes.NewHTTPError(http.StatusForbidden, "cannot process message to mismatch member")
+	}
 	if m.Type == raftpb.MsgApp {
 		s.stats.RecvAppendReq(types.ID(m.From).String(), m.Size())
 	}

--- a/server/etcdserver/server_test.go
+++ b/server/etcdserver/server_test.go
@@ -470,7 +470,7 @@ func TestApplyConfigChangeUpdatesConsistIndex(t *testing.T) {
 		lgMu:         new(sync.RWMutex),
 		lg:           lg,
 		memberId:     1,
-		r:            *realisticRaftNode(lg),
+		r:            *realisticRaftNode(lg, nil),
 		cluster:      cl,
 		w:            wait.New(),
 		consistIndex: ci,
@@ -514,9 +514,15 @@ func TestApplyConfigChangeUpdatesConsistIndex(t *testing.T) {
 	assert.Equal(t, consistIndex, rindex)
 }
 
-func realisticRaftNode(lg *zap.Logger) *raftNode {
+func realisticRaftNode(lg *zap.Logger, snap *raftpb.Snapshot) *raftNode {
 	storage := raft.NewMemoryStorage()
 	storage.SetHardState(raftpb.HardState{Commit: 0, Term: 0})
+	if snap != nil {
+		err := storage.ApplySnapshot(*snap)
+		if err != nil {
+			panic(err)
+		}
+	}
 	c := &raft.Config{
 		ID:              1,
 		ElectionTick:    10,
@@ -886,6 +892,56 @@ func TestAddMember(t *testing.T) {
 	}
 	if cl.Member(1234) == nil {
 		t.Errorf("member with id 1234 is not added")
+	}
+}
+
+// TestProcessIgnoreMismatchMessage tests Process must ignore messages to
+// mismatch member.
+func TestProcessIgnoreMismatchMessage(t *testing.T) {
+	lg := zaptest.NewLogger(t)
+	cl := newTestCluster(t)
+	st := v2store.New()
+	cl.SetStore(st)
+	be, _ := betesting.NewDefaultTmpBackend(t)
+	defer betesting.Close(t, be)
+	cl.SetBackend(schema.NewMembershipBackend(lg, be))
+
+	// Bootstrap a 3-node cluster
+	cl.AddMember(&membership.Member{ID: types.ID(1)}, true)
+	cl.AddMember(&membership.Member{ID: types.ID(2)}, true)
+	cl.AddMember(&membership.Member{ID: types.ID(3)}, true)
+	r := realisticRaftNode(lg, &raftpb.Snapshot{
+		Metadata: raftpb.SnapshotMetadata{
+			Index:     11, // magic number
+			Term:      11, // magic number
+			ConfState: raftpb.ConfState{Voters: []uint64{1, 2, 3}},
+		},
+	})
+	s := &EtcdServer{
+		lgMu:         new(sync.RWMutex),
+		lg:           lg,
+		r:            *r,
+		v2store:      st,
+		cluster:      cl,
+		reqIDGen:     idutil.NewGenerator(0, time.Time{}),
+		SyncTicker:   &time.Ticker{},
+		consistIndex: cindex.NewFakeConsistentIndex(0),
+		beHooks:      serverstorage.NewBackendHooks(lg, nil),
+	}
+	// Mock a mad switch dispatching messages to wrong node.
+	m := raftpb.Message{
+		Type:   raftpb.MsgHeartbeat,
+		To:     2, // Wrong ID.
+		From:   3,
+		Term:   11,
+		Commit: 42, // Commit is larger than the last index.
+	}
+	if types.ID(m.To) == s.MemberId() {
+		t.Fatalf("To must mismatch")
+	}
+	err := s.Process(context.Background(), m)
+	if err == nil {
+		t.Fatalf("Must ignore the message and return an error")
 	}
 }
 

--- a/server/etcdserver/server_test.go
+++ b/server/etcdserver/server_test.go
@@ -470,7 +470,7 @@ func TestApplyConfigChangeUpdatesConsistIndex(t *testing.T) {
 		lgMu:         new(sync.RWMutex),
 		lg:           lg,
 		memberId:     1,
-		r:            *realisticRaftNode(lg, nil),
+		r:            *realisticRaftNode(lg, 1, nil),
 		cluster:      cl,
 		w:            wait.New(),
 		consistIndex: ci,
@@ -514,7 +514,7 @@ func TestApplyConfigChangeUpdatesConsistIndex(t *testing.T) {
 	assert.Equal(t, consistIndex, rindex)
 }
 
-func realisticRaftNode(lg *zap.Logger, snap *raftpb.Snapshot) *raftNode {
+func realisticRaftNode(lg *zap.Logger, id uint64, snap *raftpb.Snapshot) *raftNode {
 	storage := raft.NewMemoryStorage()
 	storage.SetHardState(raftpb.HardState{Commit: 0, Term: 0})
 	if snap != nil {
@@ -524,7 +524,7 @@ func realisticRaftNode(lg *zap.Logger, snap *raftpb.Snapshot) *raftNode {
 		}
 	}
 	c := &raft.Config{
-		ID:              1,
+		ID:              id,
 		ElectionTick:    10,
 		HeartbeatTick:   1,
 		Storage:         storage,
@@ -906,20 +906,25 @@ func TestProcessIgnoreMismatchMessage(t *testing.T) {
 	defer betesting.Close(t, be)
 	cl.SetBackend(schema.NewMembershipBackend(lg, be))
 
-	// Bootstrap a 3-node cluster
+	// Bootstrap a 3-node cluster, member IDs: 1 2 3.
 	cl.AddMember(&membership.Member{ID: types.ID(1)}, true)
 	cl.AddMember(&membership.Member{ID: types.ID(2)}, true)
 	cl.AddMember(&membership.Member{ID: types.ID(3)}, true)
-	r := realisticRaftNode(lg, &raftpb.Snapshot{
+	// r is initialized with ID 1.
+	r := realisticRaftNode(lg, 1, &raftpb.Snapshot{
 		Metadata: raftpb.SnapshotMetadata{
-			Index:     11, // magic number
-			Term:      11, // magic number
-			ConfState: raftpb.ConfState{Voters: []uint64{1, 2, 3}},
+			Index: 11, // Magic number.
+			Term:  11, // Magic number.
+			ConfState: raftpb.ConfState{
+				// Member ID list.
+				Voters: []uint64{1, 2, 3},
+			},
 		},
 	})
 	s := &EtcdServer{
 		lgMu:         new(sync.RWMutex),
 		lg:           lg,
+		memberId:     1,
 		r:            *r,
 		v2store:      st,
 		cluster:      cl,
@@ -931,13 +936,13 @@ func TestProcessIgnoreMismatchMessage(t *testing.T) {
 	// Mock a mad switch dispatching messages to wrong node.
 	m := raftpb.Message{
 		Type:   raftpb.MsgHeartbeat,
-		To:     2, // Wrong ID.
+		To:     2, // Wrong ID, s.MemberId() is 1.
 		From:   3,
 		Term:   11,
-		Commit: 42, // Commit is larger than the last index.
+		Commit: 42, // Commit is larger than the last index 11.
 	}
 	if types.ID(m.To) == s.MemberId() {
-		t.Fatalf("To must mismatch")
+		t.Fatalf("m.To (%d) is expected to mismatch s.MemberId (%d)", m.To, s.MemberId())
 	}
 	err := s.Process(context.Background(), m)
 	if err == nil {

